src/Features/Core/Portable/FindReferences/IDefinitionsAndReferencesFactory.cs

```diff
@@ -0,0 +1,273 @@
+            if (thirdPartyItem != null)
+            {
+                definitions.Add(thirdPartyItem);
+            }
+        }
+
+        /// <summary>
+        /// Provides an extension point that allows for other workspace layers to add additional
+        /// results to the results found by the FindReferences engine.
+        /// </summary>
+        protected virtual DefinitionItem GetThirdPartyDefinitionItem(
+            Solution solution, ISymbol definition)
+        {
+            return null;
+        }
+
+        private static DefinitionItem CreateDefinitionItem(
+            Solution solution, ReferencedSymbol referencedSymbol)
+        {
+            var definition = referencedSymbol.Definition;
+
+            // First, determine the set of locations for this symbol that we'll
+            // want to present to the user.  Some symbols (like namespace) may
+            // have many locations, but we'll only want to show the user a single
+            // one.
+            var filteredLocations = FilterDefinitionLocations(definition);
+
+            // Convert the filtered set of locations to DefinitionLocation instances.
+            // If weren't any to, say because the symbol had no locations that we
+            // could understand (like Location.None), then we skip this item.
+            var definitionLocations = ConvertLocations(solution, referencedSymbol, filteredLocations);
+            var displayParts = definition.ToDisplayParts(s_definitionDisplayFormat).ToTaggedText();
+
+            return new DefinitionItem(
+                GlyphTags.GetTags(definition.GetGlyph()),
+                displayParts,
+                definitionLocations,
+                definition.ShouldShowWithNoReferenceLocations());
+        }
+
+        private static ImmutableArray<DefinitionLocation> ConvertLocations(
+            Solution solution, ReferencedSymbol referencedSymbol, ImmutableArray<Location> locations)
+        {
+            var definition = referencedSymbol.Definition;
+            var result = ImmutableArray.CreateBuilder<DefinitionLocation>();
+
+            // If it's a namespace, don't create any normal lcoation.  Namespaces
+            // come from many different sources, but we'll only show a single 
+            // root definition node for it.  That node won't be navigable.
+            if (definition.Kind != SymbolKind.Namespace)
+            {
+                foreach (var location in locations)
+                {
+                    if (location.IsInMetadata)
+                    {
+                        var firstSourceReferenceLocation = referencedSymbol.Locations.FirstOrNullable();
+                        if (firstSourceReferenceLocation != null)
+                        {
+                            result.Add(DefinitionLocation.CreateSymbolLocation(
+                                definition, firstSourceReferenceLocation.Value.Document.Project));
```

Comment:
No, it's actually fine (though confusing).  We're not going to store the actual symbol in the DefinitionLocation.  That would hold onto too much stuff.  Instead we just hold onto a SymbolKey so that we can resolve the symbol later on.  Because we know the symbol was referenced in this project, then we know that this project is perfectly fine to do the resolution in (otherwise the project couldn't have referenced the symbol in the first place).